File: android/core/common/src/main/kotlin/com/nexussoft/verbum/common/BookMatcher.kt
package com.nexussoft.verbum.common

import com.nexussoft.verbum.models.BibleBook
import com.nexussoft.verbum.models.BookLanguage
import com.nexussoft.verbum.models.localizedName
import com.nexussoft.verbum.models.localizedAbbreviations
import java.text.Normalizer
import java.util.Locale

/**
 * Finds books a partial query could mean: `sam` → 1 Samuel, 2 Samuel; `jn` → John;
 * `song` → Song of Solomon. Pure, deterministic, canon-ordered. Mirrors iOS.
 */
object BookMatcher {
    /**
     * Books whose name, any word of the name, or any abbreviation starts with the query
     * (case-insensitive, periods and spaces ignored). Books the query names exactly come
     * first (`jn` → John before Jonah); within each group, canon order. Empty query → nothing.
     */
    fun books(matching: String, language: BookLanguage = BookLanguage.ENGLISH): List<BibleBook> {
        val key = normalize(matching)
        if (key.isEmpty()) return emptyList()
        val exact = mutableListOf<BibleBook>()
        val prefix = mutableListOf<BibleBook>()
        for (book in BibleBook.canon) {
            val keys = listOf(book.name, book.localizedName(language)) + book.abbreviations + book.localizedAbbreviations(language)
            when {
                keys.any { normalize(it) == key } -> exact += book
                keys.any { normalize(it).startsWith(key) } ||
                    listOf(book.name, book.localizedName(language)).flatMap { it.split(" ") }.any { word -> normalize(word).startsWith(key) && !word.all(Char::isDigit) } -> prefix += book
            }
        }
        return exact + prefix
    }

    internal fun normalize(text: String): String = Normalizer.normalize(text, Normalizer.Form.NFD).lowercase(Locale.ROOT).filterNot { Character.getType(it) == Character.NON_SPACING_MARK.toInt() || it == '.' || it.isWhitespace() }
}

File: VerbumKit/Sources/Core/BookMatcher.swift
import Foundation
import Models

/// Finds books a partial query could mean: `sam` → 1 Samuel, 2 Samuel;
/// `jn` → John; `song` → Song of Solomon. Pure, deterministic, canon-ordered.
public enum BookMatcher {
    /// Books whose name, any word of the name, or any abbreviation starts with
    /// the query (case-insensitive, periods and spaces ignored). Books the
    /// query names exactly come first (`jn` → John before Jonah); within each
    /// group, canon order. Empty query matches nothing.
    public static func books(matching query: String, language: BookLanguage = .english) -> [BibleBook] {
        let key = normalize(query)
        guard !key.isEmpty else { return [] }
        var exact: [BibleBook] = []
        var prefix: [BibleBook] = []
        for book in BibleBook.canon {
            let keys = [book.name, book.localizedName(for: language)] + book.abbreviations + book.localizedAbbreviations(for: language)
            if keys.contains(where: { normalize($0) == key }) {
                exact.append(book)
            } else if keys.contains(where: { normalize($0).hasPrefix(key) })
                || [book.name, book.localizedName(for: language)].flatMap({ $0.split(separator: " ") }).contains(where: { normalize(String($0)).hasPrefix(key) && !$0.allSatisfy(\.isNumber) }) {
                prefix.append(book)
            }
        }
        return exact + prefix
    }

    static func normalize(_ text: String) -> String {
        text.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: Locale(identifier: "en_US_POSIX")).filter { $0 != "." && !$0.isWhitespace }
    }
}
